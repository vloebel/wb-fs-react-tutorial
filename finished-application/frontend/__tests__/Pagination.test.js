import { render } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import Pagination, { PAGINATION_QUERY } from '../components/Pagination';
import { makePaginationMocksFor } from '../lib/testUtils';

describe('<Pagination/>', () => {
  it('displays a loading message', async () => {
    const { container, queryByText } = render(
      <MockedProvider mocks={makePaginationMocksFor(1)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    expect(queryByText('Loading...')).toBeTruthy();
  });

  it('renders pagination for 18 items', async () => {
    const { container, findByTestId, getByTestId, debug, query } = render(
      <MockedProvider mocks={makePaginationMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    // wait for pagination to get past loading
    await findByTestId('pagination');
    expect(getByTestId('totalPages')).toHaveTextContent('5');
    expect(container).toMatchSnapshot();
  });

  it('disables prev button on first page', async () => {
    const { getByText, findByTestId } = render(
      <MockedProvider mocks={makePaginationMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await findByTestId('pagination');
    expect(getByText(/Prev/)).toHaveAttribute('aria-disabled', 'true');
    expect(getByText(/Next/)).toHaveAttribute('aria-disabled', 'false');
  });

  it('disables next button on last page', async () => {
    const { getByText, findByTestId } = render(
      <MockedProvider mocks={makePaginationMocksFor(18)}>
        <Pagination page={5} />
      </MockedProvider>
    );
    await findByTestId('pagination');
    expect(getByText(/Prev/)).toHaveAttribute('aria-disabled', 'false');
    expect(getByText(/Next/)).toHaveAttribute('aria-disabled', 'true');
  });

  it('enables all buttons on a middle page', async () => {
    const { getByText, findByTestId } = render(
      <MockedProvider mocks={makePaginationMocksFor(18)}>
        <Pagination page={3} />
      </MockedProvider>
    );
    await findByTestId('pagination');
    expect(getByText(/Prev/)).toHaveAttribute('aria-disabled', 'false');
    expect(getByText(/Next/)).toHaveAttribute('aria-disabled', 'false');
  });
});
