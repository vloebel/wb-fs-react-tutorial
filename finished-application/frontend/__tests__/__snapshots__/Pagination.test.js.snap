// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> renders pagination for 18 items 1`] = `
<div>
  <div
    class="PaginationStyles-aduuar-0 lmlpoS"
    data-testid="pagination"
  >
    <a
      aria-disabled="true"
      class="prev"
      href="/products/0"
    >
      ← Prev
    </a>
    <p>
      Page 
      1
       of
       
      <span
        class="totalPages"
        data-testid="totalPages"
      >
        5
      </span>
    </p>
    <p>
      18
       Items Total
    </p>
    <a
      aria-disabled="false"
      class="next"
      href="/products/2"
    >
      Next →
    </a>
  </div>
</div>
`;
